python/ray/data/_internal/execution/resource_manager.py

```diff
@@ -220,81 +269,288 @@ class ReservationOpResourceLimiter(OpResourceLimiter):
+            if idle_time < cls.WARN_ON_IDLE_TIME_S or cls._warn_printed:
+                return
+            cls._warn_printed = True
+            msg = (
+                f"Operator {op} is running but has no outputs for {idle_time} seconds."
+                " Execution may be slower than expected.\n"
+                "Ignore this warning if your UDF is expected to be slow."
+                " Otherwise, this can happen when there are fewer cluster resources"
+                " available to Ray Data than expected."
+                " If you have non-Data tasks or actors running in the cluster, exclude"
+                " their resources from Ray Data with"
+                " `DataContext.get_current().execution_options.exclude_resources`."
+                " This message will only print once."
+            )
+            logger.get_logger(log_to_stdout=True).warning(msg)
+
     def __init__(self, resource_manager: ResourceManager, reservation_ratio: float):
         super().__init__(resource_manager)
         self._reservation_ratio = reservation_ratio
         assert 0.0 <= self._reservation_ratio <= 1.0
-        # We only limit map operators.
-        self._eligible_ops = [
-            op for op in self._resource_manager._topology if isinstance(op, MapOperator)
-        ]
-        # Per-op reserved resources.
+        # Per-op reserved resources, excluding `_reserved_for_op_outputs`.
         self._op_reserved: Dict[PhysicalOperator, ExecutionResources] = {}
+        # Memory reserved for the outputs of each operator, including the operator's
+        # internal and external output buffers and next operator's input buffers, but
+        # not including the pending task outputs.
+        # Note, if we don't reserve memory for outputs, all the budget may be used by
+        # by the pending task outputs. Then we'll have no budget to pull the outputs
+        # from the running tasks.
+        self._reserved_for_op_outputs: Dict[PhysicalOperator, int] = {}
         # Total shared resources.
         self._total_shared = ExecutionResources.zero()
-        # Resource limits for each operator.
-        self._op_limits: Dict[PhysicalOperator, ExecutionResources] = {}
+        # Resource budgets for each operator.
+        self._op_budgets: Dict[PhysicalOperator, ExecutionResources] = {}
+        # Whether each operator has reserved the minimum resources to run
+        # at least one task.
+        # This is used to avoid edge cases where the entire resource limits are not
+        # enough to run one task of each op.
+        # See `test_no_deadlock_on_small_cluster_resources` as an example.
+        self._reserved_min_resources: Dict[PhysicalOperator, bool] = {}
+
         self._cached_global_limits = ExecutionResources.zero()
+        self._cached_num_eligible_ops = 0
+
+        self._idle_detector = self.IdleDetector()
+
+    def _get_eligible_ops(self) -> List[PhysicalOperator]:
```

Comment:
We should only enable the feature when the plan only has MapOperator, right? Have we addressed the comment to disable the feature otherwise - https://github.com/ray-project/ray/pull/43026#discussion_r1487002159 ?

Reply:
it will only be enabled when the dateset has [these ops](https://github.com/ray-project/ray/blob/1c2b777e242099c1af92200ec9ad112ec6eb82fd/python/ray/data/_internal/execution/resource_manager.py#L45)